eclair-core/src/main/scala/fr/acinq/eclair/router/Validation.scala

```diff
@@ -414,51 +415,58 @@ object Validation {
-                // we already have a pending request to that node, let's add this channel to the list and we'll get it later
-                d.copy(sync = d.sync + (remoteNodeId -> sync.copy(remainingQueries = sync.remainingQueries :+ query, totalQueries = sync.totalQueries + 1)))
-              case _ =>
-                // otherwise we send the query right away
-                peerConnection ! query
-                d.copy(sync = d.sync + (remoteNodeId -> Syncing(remainingQueries = Nil, totalQueries = 1)))
-            }
-          case _ =>
-            // we don't know which node this update came from (maybe it was stashed and the channel got pruned in the meantime or some other corner case).
-            // or we don't have a peerConnection to send our query to.
-            // anyway, that's not really a big deal because we have removed the channel from the pruned db so next time it shows up we will revalidate it
-            d
+          d.prunedChannels.get(RealShortChannelId(u.shortChannelId.toLong)) match {
+            case Some(pc) =>
+              // The channel was pruned (one of the two channel updates was older than 2 weeks).
+              val pc1 = pc.applyChannelUpdate(update)
+              if (!Announcements.checkSig(u, pc.getNodeIdSameSideAs(u))) {
+                log.warning("bad signature for announcement shortChannelId={} {}", u.shortChannelId, u)
+                sendDecision(origins, GossipDecision.InvalidSignature(u))
+                d
+              } else if (pc1.isStale(currentBlockHeight)) {
+                log.debug("channel shortChannelId={} is still stale and should stay pruned", u.shortChannelId)
+                sendDecision(origins, GossipDecision.RelatedChannelPruned(u))
+                // We update our state with this new channel update.
+                db.updateChannel(u)
+                val prunedChannels1 = d.prunedChannels + (pc1.shortChannelId -> pc1)
+                d.copy(prunedChannels = prunedChannels1)
+              } else {
+                log.info("channel shortChannelId={} is back from the dead", u.shortChannelId)
+                // We notify front nodes that the channel is back.
+                ctx.system.eventStream.publish(ChannelsDiscovered(SingleChannelDiscovered(pc1.ann, pc1.capacity, pc1.update_1_opt, pc1.update_2_opt) :: Nil))
+                // We move the channel out of the pruned map and re-validate updates as if we were receiving them for the first time.
+                val d1 = d.copy(channels = d.channels + (pc.shortChannelId -> pc.copy(update_1_opt = None, update_2_opt = None)), prunedChannels = d.prunedChannels - pc.shortChannelId)
+                (pc1.update_1_opt ++ pc1.update_2_opt).foldLeft(d1) {
+                  case (current, chanUpdate) =>
+                    if (chanUpdate.channelFlags.isNode1 == u.channelFlags.isNode1) {
+                      handleChannelUpdate(current, db, currentBlockHeight, update)
```

Comment:
It feels strange to see `handleChannelUpdate` call itself here ? Wouldn't be safer (and easier to parse) to extract code that needs to be used here and call it explicitly ?

Reply:
I don't feel too strongly about this but `handleChannelUpdate` has many side-effects and it's a bit difficult to reason about what happens when it calls itself again, but I also see why it's the simplest way to not forget to apply one of these side-effects.

Reply:
But it should be possible to extract what should be done in each case move it out of the match/case block ?

Reply:
I removed usage of the existing `addPublicChannel` and `handleChannelUpdate` methods that had side-effects in https://github.com/ACINQ/eclair/pull/2456/commits/47f8f97e9558bc947733f7aa2aa71d805c0ce45f